### ISSUE TITLE
CRAN check failures with some LAPACK versions

### ISSUE THREAD

**Issue by bbolker** [lme4/lme4#677]
According to CRAN maintainers (specifically Kurt Hornik), newly released version of lme4 is failing on platforms {r-devel-linux-x86_64-debian-gcc ;  r-devel-linux-x86_64-debian-clang ; r-oldrel-windows-ix86+x86_64 }. Kurt attributes this to

> The failing check runs are using an external LAPACK 3.10.0. ... To be even more precise, Brian does not get this on Fedora with
a system reference (netlib) 3.10 LAPACK, whereas I do on Debian with
OpenBLAS.

The failing test appears to be [here](https://github.com/lme4/lme4/blob/master/tests/testthat/test-lmer.R#L224-L227=): minimal replication *should* be:

```r
library(lme4); library(testthat)
d <- data.frame(z=rnorm(200),
                    f=factor(sample(1:10,200, replace=TRUE)))
## test messed-up Hessian
fm1 <- lmer(z~ as.numeric(f) + 1|f, d)
fm1@optinfo$derivs$Hessian[2,2] <- NA
expect_warning(lme4:::checkConv(fm1@optinfo$derivs,
                     coefs=c(1,1),
                     ctrl=lmerControl()$checkConv,lbound=0),
                   "Problem with Hessian check")
```

and the proximal point of failure seems to be [here](https://github.com/lme4/lme4/blob/master/R/checkConv.R#L63-L69=): looking at the logic, I presume that with whatever version of LAPACK is being used, `solve(chol(Hessian),gradient))` is returning `NA` but **not** throwing an error (I get "Error in chol.default(derivs$Hessian) :  the leading minor of order 2 is not positive definite").

I could probably fix this by changing the test to:

```r 
if (inherits(scgrad, "error") || any(is.na(scgrad))) {
```

Frustratingly, I haven't been able to come up with a working test platform for this -- I've spent a few hours screwing around with Docker/rocker-project containers, without any luck. 

**update**: have spent all afternoon going around in LAPACK/BLAS circles. See notes [here](https://github.com/lme4/lme4/blob/master/misc/R_lapackstuff.md)

```
  == Failed tests ================================================================
  -- Error (test-lmer.R:224:5): lmer ---------------------------------------------
  Error in `if (maxmingrad > ccl$tol) {
      w <- which.max(maxmingrad)
      res$code <- -1L
      wstr <- gettextf("Model failed to converge with max|grad| = %g (tol = %g, component %d)",
          maxmingrad, ccl$tol, w)
  }`: missing value where TRUE/FALSE needed
  Backtrace:
      x
   1. +-testthat::expect_warning(...) at test-lmer.R:224:4
   2. | \-testthat:::quasi_capture(enquo(object), label, capture_warnings)
   3. | +-testthat .capture(...)
   4. | | \-base::withCallingHandlers(...)
   5. | \-rlang::eval_bare(quo_get_expr(.quo), quo_get_env(.quo))
   6. \-lme4:::checkConv(...)
 ```

**Comment by mmaechler** [comment]
I'm still puzzled by the message `missing value where TRUE/FALSE needed`   which really  says that one of the two 
`maximingrad`  and `ccl$tol` is NA  .... and I have not seen you addressing this fact ... but then I have not yet delved deep down into `checkConv()` ... hopefully will get to this later or tomorrow.

Not having testthat  would really also make things more clear (of where to look)

**Comment by bbolker** [comment]
My logic is as follows: 

- `NA` in `if` statement
- implies (as you said) `maxmingrad` and/or `ccl$tol` are `NA`, I think it's `maxmingrad`
- therefore `mingrad` contains an `NA`
- therefore either `scgrad` or `derivs$gradient` contains an `NA`, I'm sure it's `scgrad`
- therefore I'm concluding that `with(derivs, solve(chol(Hessian), gradient))` returned an answer containing an `NA` rather than throwing an error

```
scgrad <- tryCatch(with(derivs,solve(chol(Hessian),gradient)),
                       error=function(e)e)
if (inherits(scgrad, "error")) {
        wstr <- "unable to evaluate scaled gradient"
        res$code <- -1L
} else {
       mingrad <- pmin(abs(scgrad),abs(derivs$gradient))
       maxmingrad <- max(mingrad)
       if (maxmingrad > ccl$tol) {  ## ERROR here
            ...
       }
}
```

**Comment by mmaechler** [comment]
Thank you.   That makes sense.
Still we need a platform.    Dirk Eddelbuettel (<-> "rocker") would have r-devel docker containers with Debian  and we could ask for help either from him or the CRAN team, i.e. Kurt  on how to make sure to get an *external* Lapack  3.1.0

**Comment by bbolker** [comment]
Closed (I think!) by 58df23967 (without ever having managed to actually reproduce the problem).
